File: LocationManagerObjC/LocationManagerObjC/Location/LocationManager.m
//
//  LocationManager.m
//  LocationManagerObjC
//
//  Created by Jason Lew on 6/11/16.
//  Copyright © 2016 Jason Lew. All rights reserved.
//

#import "LocationManager.h"
#import "JALLog.h"

@interface LocationManager()<CLLocationManagerDelegate>

@property (strong, nonatomic) CLLocationManager *locationManager;
@property (strong, nonatomic) CLLocation *bestEffortLocation;

@end
@implementation LocationManager

- (instancetype)init {
    if (self = [super init]) {
        _locationManager = [[CLLocationManager alloc]init];
        _locationManager.delegate = self;
        _locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters;
        _locationManager.pausesLocationUpdatesAutomatically = YES;
        _locationManager.activityType = CLActivityTypeOther;
        [self checkLocationAuthorizationStatus];
        return self;
    }
    return nil;
}

- (void)dealloc
{
    if ([CLLocationManager locationServicesEnabled]) {
        [self.locationManager stopUpdatingLocation];
    }
    
    self.bestEffortLocation = nil;
}

#pragma mark - Actions

- (void)startLocationServices {
    if ([CLLocationManager locationServicesEnabled]) {
        [self.locationManager requestWhenInUseAuthorization];
    }
}

// Be sure to add NSLocationWhenInUseUsageDescription and a string message in info.plist
- (void)checkLocationAuthorizationStatus {
    switch ([CLLocationManager authorizationStatus]) {
        case kCLAuthorizationStatusNotDetermined:
            [self startLocationServices];
            DLog(@"Starting location services.");
            break;
        case kCLAuthorizationStatusDenied:
        case kCLAuthorizationStatusRestricted:
            DLog(@"Location services not available.");
            break;
        case kCLAuthorizationStatusAuthorizedWhenInUse:
        case kCLAuthorizationStatusAuthorizedAlways:
            [self.locationManager startUpdatingLocation];
            break;
        default:
            break;
    }
}

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    switch (status) {
        case kCLAuthorizationStatusAuthorizedWhenInUse:
        case kCLAuthorizationStatusAuthorizedAlways:
            [self.locationManager startUpdatingLocation];
            break;
        case kCLAuthorizationStatusDenied:
        case kCLAuthorizationStatusRestricted:
            [self.locationManager stopUpdatingLocation];
            break;
        case kCLAuthorizationStatusNotDetermined:
            [self startLocationServices];
            break;
        default:
            break;
    }
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    if (!locations.lastObject) { return; };
    CLLocation *location = locations.lastObject;
    // Test that this isn't an invalid measurement
    if (location.horizontalAccuracy < 0) {
        self.bestEffortLocation = nil;
        return;
    };
    
    // Test if the location is cached
    NSTimeInterval locationAge = -(location.timestamp.timeIntervalSinceNow);
    if (locationAge > 5.0) {
        self.bestEffortLocation = nil;
        return;
    }
    
    // Test if the new location is more accurate
    if (!self.bestEffortLocation || self.bestEffortLocation.horizontalAccuracy > location.horizontalAccuracy) {
        self.bestEffortLocation = location;
        
        [self.delegate foundInitialLocation:location];
        
        // Test if it meets the desired accuracy
        if (location.horizontalAccuracy <= self.locationManager.desiredAccuracy) {
            // The measurement staisfies the requirement.
            [self.delegate bestEffortLocationFound:location];
        }
        DLog(@"Best effort location: %@", self.bestEffortLocation);
    }
}

@end

File: LocationManagerSwift/LocationManagerSwift/LocationManager.swift
//
//  LocationManager.swift
//  LocationManagerSwift
//
//  Created by Jason Lew on 5/21/16.
//  Copyright © 2016 Jason Lew. All rights reserved.
//

import Foundation
import CoreLocation

protocol LocationManagerDelegate: class {
    func bestEffortLocationFound(location: CLLocation)
    func foundInitialLocation(location: CLLocation)
}

class LocationManager: NSObject, CLLocationManagerDelegate {
    
    struct LocationTimeInterval {
        static let timeout = NSTimeInterval(30)
        static let restartAfter = NSTimeInterval(60)
    }
    
    // MARK: - Properties
    weak var delegate: LocationManagerDelegate?
    
    lazy var locationManager = CLLocationManager()
    var bestEffortAtLocation: CLLocation?
    var timer: NSTimer?
    
    // MARK: - Init
    init(locationAccuracy: CLLocationAccuracy? = nil) {
        super.init()
        locationManager.delegate = self
        locationManager.desiredAccuracy =
            locationAccuracy != nil ? locationAccuracy! : kCLLocationAccuracyNearestTenMeters
        locationManager.activityType = .Other
        locationManager.pausesLocationUpdatesAutomatically = true
        checkLocationAuthorizationStatus()
    }
    
    deinit {
        if CLLocationManager.locationServicesEnabled() {
            locationManager.stopUpdatingLocation()
        }
        
        if timer != nil {
            cancelTimer()
        }
        
        bestEffortAtLocation = nil
    }
    
    // MARK: - Actions
    
    func cancelTimer() {
        timer?.invalidate()
        timer = nil
    }
    
    func startLocationServices() {
        if CLLocationManager.locationServicesEnabled() {
            locationManager.requestWhenInUseAuthorization()
        }
    }
    
    func startUpdatingLocation() {
        DLog.print("Updating location.")
        locationManager.startUpdatingLocation()
        
        // Stop the Core Location Manager after delay
        cancelTimer()
        timer = NSTimer.scheduledTimerWithTimeInterval(
            LocationTimeInterval.timeout,
            target: self,
            selector: #selector(LocationManager.stopUpdatingLocationWithDelayedRestart),
            userInfo: nil, repeats: false)
    }
    
    func stopUpdatingLocationWithDelayedRestart() {
        // The location update is suspended to limit power consumption
        locationManager.stopUpdatingLocation()
        cancelTimer()
        timer = NSTimer.scheduledTimerWithTimeInterval(
            LocationTimeInterval.restartAfter,
            target: self,
            selector: #selector(LocationManager.startUpdatingLocation),
            userInfo: nil, repeats: false)
    }
    
    // Add NSLocationWhenInUseUsageDescription to info.plist
    func checkLocationAuthorizationStatus() {
        switch CLLocationManager.authorizationStatus() {
        case .NotDetermined:
            startLocationServices()
            DLog.print("Starting location services.")
        case .Denied, .Restricted:
            DLog.print("Location services not available.")
        case .AuthorizedWhenInUse, .AuthorizedAlways:
            startUpdatingLocation()
        }
    }
    
    func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
        switch status {
        case .AuthorizedWhenInUse, .AuthorizedAlways:
            locationManager.startUpdatingLocation()
        case .Denied, .Restricted:
            locationManager.stopUpdatingLocation()
        case .NotDetermined:
            startLocationServices()
        }
    }
    
    func locationManager(manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else {
            DLog.print("No valid location")
            return
        }
        // Test that it isn't an invalid measurement
        if location.horizontalAccuracy < 0 {
            bestEffortAtLocation = nil
            return
        }
        
        // Test if the location is cached
        let locationAge = -(location.timestamp.timeIntervalSinceNow)
        if locationAge > 5.0 {
            bestEffortAtLocation = nil
            return
        }
        
        // Test if the new location is more accurate
        if bestEffortAtLocation == nil || bestEffortAtLocation?.horizontalAccuracy > location.horizontalAccuracy {
            bestEffortAtLocation = location
            
            delegate?.foundInitialLocation(location)
            
            // Test if it meets the desired accuracy
            if location.horizontalAccuracy <= locationManager.desiredAccuracy {
                // The measurement satisfies the requirement.
                if let bestEffortLocation = bestEffortAtLocation {
                    delegate?.bestEffortLocationFound(bestEffortLocation)
                }
                stopUpdatingLocationWithDelayedRestart()
                DLog.print("Best effort location: \(bestEffortAtLocation)")
            }
        }
    }
}
